File: src/trees/level_order_traversal.py
from collections import deque
from typing import Optional, List

from src.trees.utils import TreeNode


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        levels = []
        queue = deque()
        queue.append(root)

        while queue:
            level = []
            levelSize = len(queue)
            for i in range(levelSize):
                node: TreeNode = queue.popleft()
                if node:
                    level.append(node.val)
                    queue.append(node.left)
                    queue.append(node.right)
            if level:
                levels.append(level)

        return levels

File: java/src/main/java/com/blind75/trees/LevelOrderTraversal.java
package com.blind75.trees;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;

public class LevelOrderTraversal {
    public List<List<Integer>> levelOrder(TreeNode treeNode) {
        List<List<Integer>> levels = new ArrayList<>();
        var queue = new ArrayDeque<TreeNode>();
        if (treeNode != null) {
            queue.add(treeNode);
        }

        while (!queue.isEmpty()) {
            List<Integer> level = new ArrayList<>();
            var queueSize = queue.size();
            for (int i = 0; i < queueSize; i++) {
                var node = queue.remove();
                if (node != null) {
                    level.add(node.val);
                }
                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null) {
                    queue.add(node.right);
                }
            }
            if (!level.isEmpty()) {
                levels.add(level);
            }
        }
        return levels;
    }
}
